chunky/src/java/se/llbit/chunky/renderer/scene/SkyCache.java

```diff
@@ -0,0 +1,162 @@
+  private int skyResolution = 1024;
+
+  private SimulatedSky simSky;
+  private Sky sky;
+
+  private int version = 0;
+
+  /**
+   * An on-the-fly sky cache. Automatically calculates sky colors as they are requested and caches them. Any repeat
+   * requests will pull from the cache.
+   *
+   * Default cache size is 1024x1024 with bilinear interpolation which seems to work well for all simulated sky
+   * modes.
+   *
+   * @param sky Sky object to pull sky renderer from.
+   */
+  public SkyCache(Sky sky) {
+    this.sky = sky;
+    skyTexture = null;
+  }
+
+  /** Sync this cache with another cache */
+  public void syncCache(SkyCache cache) {
+    if (this.skyResolution != cache.skyResolution) {
+      setSkyResolution(cache.skyResolution);
+    }
+
+    if (this.simSky != cache.simSky) {
+      this.simSky = cache.simSky;
+      skyTexture = null;
+    }
+
+    if (this.version != cache.version) {
+      this.version = cache.version;
+      skyTexture = null;
+    }
+  }
+
+  /** Reset the sky cache */
+  public void reset(Sky sky) {
+    simSky = sky.getSimulatedSky();
+    version += 1;
+
+    // Sky has not yet been initialized. No need to reset.
+    if (skyTexture == null) {
+      return;
+    }
+
+    for (int i = 0; i < skyResolution+1; i++) {
+      for (int j = 0; j < skyResolution+1; j++) {
+        for (int k = 0; k < 3; k++) {
+          skyTexture[i][j][k] = -1;
+        }
+      }
+    }
+  }
+
+  /** Adjust the sky resolution and reset the cache */
+  public void setSkyResolution(int skyResolution) {
+    this.skyResolution = skyResolution;
+    skyTexture = null;
+  }
+
+  /** Get the current sky resolution */
+  public int getSkyResolution() {
+    return this.skyResolution;
+  }
+
+  /**
+   * Calculate the incident light. Will automatically pull from the cache or calculate new values. Cache values are
+   * bilinearly interpolated.
+   */
+  public Vector3 calcIncidentLight(Ray ray) {
+    if (skyTexture == null) {
+      skyTexture = new double[skyResolution+1][skyResolution+1][3];
+      reset(sky);
+    }
+
+    double theta = FastMath.atan2(ray.d.z, ray.d.x);
+    theta /= PI*2;
+    theta = ((theta % 1) + 1) % 1;
+    double phi = (FastMath.asin(QuickMath.clamp(ray.d.y, -1, 1)) + PI/2) / PI;
+
+    Vector3 color = getColorInterpolated(theta, phi);
+    ColorUtil.RGBfromHSL(color, color.x, color.y, color.z);
+    return color;
+  }
+
+  // Linear interpolation between 2 points in 1 dimension
+  private double interp1D(double x, double x0, double x1, double y0, double y1) {
+    return y0 + (x - x0)*(y1-y0)/(x1-x0);
+  }
+
+  // Calculate the bilinearly interpolated value from the cache.
+  private Vector3 getColorInterpolated(double normX, double normY) {
+    double x = normX * skyResolution;
+    double y = normY * skyResolution;
+    int floorX = (int) QuickMath.clamp(x, 0, skyResolution-1);
+    int floorY = (int) QuickMath.clamp(y, 0, skyResolution-1);
```

Comment:
Is `QuickMath` actually faster for clamping than `Math`?

Reply:
`Math` does not seem to have an explicit clamping method so you must call both `Math.min` and `Math.max`. `QuickMath` does this with some ternary operators. The difference in speed is probably negligible but `clamp` is more clear and theoretically a bit faster.

Reply:
Whoops… You're right there is no `Math.clamp`, my bad.